libsql_client/dbapi2/_utils.py

```diff
@@ -0,0 +1,199 @@
+            if pos + 1 < end_pos and sql[pos + 1] == "*":
+                pos += 2
+                # NOTE: in C pos[end_pos] == 0, so they do "pos[0]".
+                # Here we compare pos + 1 < end_pos. To avoid a messy
+                # long line, the condition is moved into an explicit "if"
+                # that "break"s.
+                while pos < end_pos:
+                    if sql[pos] == "*":
+                        if pos + 1 < end_pos and sql[pos + 1] == "/":
+                            break
+                    pos += 1
+
+                if pos >= end_pos:
+                    return None
+
+                pos += 2  # C uses for() with trailing ++
+                continue
+
+            return sql[pos:]  # not a C style comment
+
+        return sql[pos:]
+
+    return None
+
+
+_iter_sql_delim_chars = {";", ",", "(", ")", "[", "]"}
+_iter_sql_stop_chars = _lstrip_sql_whitespace_chars.union(
+    _iter_sql_delim_chars,
+)
+_iter_sql_quote_chars = {"\"", "'"}
+
+
+def _iter_sql_get_quoted_end(sql: str, pos: int, end_pos: int) -> int:
+    if pos + 1 >= end_pos:
+        return pos
+
+    escaped = False
+    ch = sql[pos]
+    pos += 1
+    while pos < end_pos:
+        if escaped:
+            escaped = False
+        else:
+            if sql[pos] == ch:
+                pos += 1
+                break
+            escaped = sql[pos] == "\\"
+        pos += 1
+
+    return pos
+
+
+def iter_sql_tokens(sql: str) -> Iterable[str]:
+    """Calls lstrip_sql() to get the start of the next token and yield it.
+
+    It will handle stop chars (;,[]()) as their own tokens, as well as handle
+    double and single quotes, as well as quote escaping with \\
+
+    >>> list(iter_sql_tokens("--COMMENT\\nBEGIN X"))
+    ['BEGIN', 'X']
+    >>> list(iter_sql_tokens(
+    ... "CREATE TABLE x (id INTEGER /* COMMENT */, name TEXT)"))
+    ['CREATE', 'TABLE', 'x', '(', 'id', 'INTEGER', ',', 'name', 'TEXT', ')']
+    >>> list(iter_sql_tokens("SELECT; INSERT; DELETE"))
+    ['SELECT', ';', 'INSERT', ';', 'DELETE']
+    >>> print("\\n".join(iter_sql_tokens(
+    ... 'INSERT INTO t (a,b) VALUES("s -- c /* c */", \\'x\\')'
+    ... )))
+    INSERT
+    INTO
+    t
+    (
+    a
+    ,
+    b
+    )
+    VALUES
+    (
+    "s -- c /* c */"
+    ,
+    'x'
+    )
+
+    """
+    while sql:
+        sql = lstrip_sql(sql) or ""
+        if not sql:
+            break
+
+        pos = 0
+        end_pos = len(sql)
+        while pos < end_pos:
+            ch = sql[pos]
+            if ch in _iter_sql_quote_chars:
+                pos = _iter_sql_get_quoted_end(sql, pos, end_pos)
+                break
+            elif ch in _iter_sql_stop_chars:
+                break
+            else:
+                pos += 1
+
+        if pos == end_pos:
+            yield sql
+            break
+
+        if pos > 0:
+            yield sql[:pos]
+            sql = sql[pos:]
+        elif sql[0] in _iter_sql_delim_chars:
```

Comment:
it's (or should be) impossible to happen

Reply:
Are you OK with falling into an infinite loop if this invariant breaks in the future? This code is very hard to follow, it will be very easy for future changes to break it.

Reply:
I can do an assert in the else clause... but it can never happen since the only way to stop at the first char (pos = 0) is when we face a _iter_sql_delim_chars

Reply:
anyway, changed to an `else`, as it should always be the case (elif condition must always be true)